### ISSUE TITLE
Panic in `PYI019` autofix

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ruff#15849]
### Description

Running `ruff check --select=PYI019 --preview foo.pyi` triggers a panic if `foo.pyi` has this contents:

```py
class F:
    @classmethod
    def m[S](cls: type[S], /) -> S[int]: ...
```

The panic occurs at this line here:

https://github.com/astral-sh/ruff/blob/59be5f5278d78e42a46b9e7c9306a93a3892ab29/crates/ruff_linter/src/rules/flake8_pyi/rules/custom_type_var_return_type.rs#L295-L297

And the reason is that, contrary to what the comment says, `uses_custom_typevar` does _not_ guarantee that the return `Expr` will be an `ExprName`, due to the `map_subscript` call here: https://github.com/astral-sh/ruff/blob/59be5f5278d78e42a46b9e7c9306a93a3892ab29/crates/ruff_linter/src/rules/flake8_pyi/rules/custom_type_var_return_type.rs#L187-L189

(The `map_subscript` call also looks buggy, FWIW.)

We should switch to a more type-safe code pattern here. Note that although the panic occurs in autofix logic, the panic occurs even if you run Ruff without `--fix`, since we create fixes for diagnostics even if `--fix` is not specified.

**Comment by AlexWaygood** [comment]
This also triggers a panic at the same location in the Rust code:

```py
class F:
    def m[S](self: S) -> S[int]: ...
```

Because there's another `map_subscript` call here, which is also incorrect:

https://github.com/astral-sh/ruff/blob/59be5f5278d78e42a46b9e7c9306a93a3892ab29/crates/ruff_linter/src/rules/flake8_pyi/rules/custom_type_var_return_type.rs#L217-L222
